wp_api/src/authenticator.rs

```diff
@@ -0,0 +1,310 @@
+use http::header::HeaderMap;
+use http::header::HeaderValue;
+use std::sync::{Arc, RwLock};
+use url::Url;
+
+use crate::{
+    request::{
+        endpoint::ApiBaseUrl, RequestExecutor, RequestMethod, WpNetworkHeaderMap, WpNetworkRequest,
+        WpNetworkRequestBody, WpNetworkResponse,
+    },
+    RequestExecutionError, WpLoginCredentials,
+};
+
+const CONTENT_TYPE_FORM: &str = "application/x-www-form-urlencoded";
+
+#[async_trait::async_trait]
+pub trait Authenticator: Send + Sync + std::fmt::Debug {
+    // The host of the site to which the authenticator is associated.
+    // This can be used to prevent credentials from being sent to other sites.
+    fn host(&self) -> &str;
+
+    async fn authentication_headers(&self) -> Option<HeaderMap>;
+
+    fn reset(&self);
+
+    fn should_authenticate(&self, request_url: &str, response_status_code: Option<u16>) -> bool;
+
+    async fn re_authenticate(
+        &self,
+        request: &WpNetworkRequest,
+        previous_response: &WpNetworkResponse,
```

Comment:
As a follow up to [this comment](https://github.com/Automattic/wordpress-rs/pull/327#discussion_r1801580794), I think we should take the `response_status_code` here.